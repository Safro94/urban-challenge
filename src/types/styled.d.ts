import 'styled-components';

interface IPalette {
	main: string;
	text: string;
}

interface IBreakpoints {
	mobile: string;
	tablet: string;
	laptop: string;
	desktop: string;
}

declare module 'styled-components' {
	export interface DefaultTheme {
		breakpoints: IBreakpoints;
		palette: {
			common: {
				background: string;
				white: string;
				black: string;
				grey: string;
			};
			primary: IPalette;
		};
	}
}
