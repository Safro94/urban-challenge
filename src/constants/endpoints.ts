export const SLOTS_ENDPOINT: string = '/slots.json';
export const WORKERS_ENDPOINT: string = '/workers.json';
export const AVAILABLE_WORKERS_ENDPOINT: string = '/available-workers.json';
