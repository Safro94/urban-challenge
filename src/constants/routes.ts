export const HOME = '/';
export const WORKERS = '/workers';
