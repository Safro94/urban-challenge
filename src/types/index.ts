export interface ISlot {
	id: number;
	localisedTime: string;
	price: string;
}

export interface IWorker {
	id: number;
	name: string;
	rating: string;
	isNew: boolean;
}

export interface ICartItem {
	id: string;
	slot: ISlot | null;
	worker: IWorker;
	price: number;
	amount: number;
}

export interface IAvailableWorker {
	slotId: number;
	availableWorkersIds: number[];
}

export interface IWorkerResponse {
	workers: IWorker[];
}

export interface ISlotsResponse {
	slots: ISlot[];
}

export interface IAvailableWorkerResponse {
	slot_id: number;
	availableWorker_ids: number[];
}

export interface IAvailableWorkersResponse {
	'available-workers': IAvailableWorkerResponse[];
}

export interface ISelectOption {
	label: string;
	value: string;
	[key: string]: unknown;
}

export enum RequestMethods {
	GET = 'get',
	POST = 'post',
	DELETE = 'delete',
}

export enum ButtonTypes {
	Primary = 'primary',
}
