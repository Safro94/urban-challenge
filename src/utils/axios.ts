import axios, { AxiosError, AxiosRequestConfig } from 'axios';
import https from 'https';

const opts: AxiosRequestConfig = {
	httpsAgent: new https.Agent({ rejectUnauthorized: false }),
	baseURL: process.env.REACT_APP_API_URL,
};

const interceptorError = (error: AxiosError) => Promise.reject(error.response);

const instance = axios.create(opts);

instance.interceptors.response.use(response => response, interceptorError);

export default instance;
