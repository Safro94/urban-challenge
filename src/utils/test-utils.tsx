import { FC, ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { BrowserRouter as Router } from 'react-router-dom';

import { theme } from '../theme';

const AllTheProviders: FC = ({ children }) => {
	return (
		<ThemeProvider theme={theme}>
			<Router>{children}</Router>
		</ThemeProvider>
	);
};

const customRender = (
	ui: ReactElement,
	options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from '@testing-library/react';
export { customRender as render };
