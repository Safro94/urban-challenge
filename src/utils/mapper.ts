import { IAvailableWorker, IAvailableWorkerResponse } from '../types';

export const mapAvailableWorkers = (
	workers: IAvailableWorkerResponse[]
): IAvailableWorker[] => {
	return workers.map((worker: IAvailableWorkerResponse) => ({
		slotId: worker.slot_id,
		availableWorkersIds: worker.availableWorker_ids,
	}));
};
