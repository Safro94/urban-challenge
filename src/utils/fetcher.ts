import { AxiosResponse } from 'axios';
import {
	IAvailableWorkersResponse,
	ISlotsResponse,
	IWorkerResponse,
	RequestMethods,
} from '../types';

import axios from './axios';

interface IFetcher {
	method?: RequestMethods;
	url: string;
	data?: null | any;
	body?: null | any;
}

type Response = IWorkerResponse | ISlotsResponse | IAvailableWorkersResponse;

const fetcher = async ({
	method = RequestMethods.GET,
	url,
	body = null,
}: IFetcher): Promise<AxiosResponse<Response>> => {
	try {
		const res: AxiosResponse<Response> = await axios[method](url, body);
		return res;
	} catch (error) {
		throw error;
	}
};

export default fetcher;
