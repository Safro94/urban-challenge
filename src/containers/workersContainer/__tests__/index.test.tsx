import userEvent from '@testing-library/user-event';

import { render, screen, waitFor } from '../../../utils/test-utils';

import WorkersContainer from '..';

import { IAvailableWorker, IWorker } from '../../../types';

import WorkerService from '../../../services/workerService';

import { HOME } from '../../../constants/routes';

jest.mock('../../../services/workerService', () => {
	return {
		getAvailableWorkers: jest.fn(),
	};
});

const mockHistoryPush = jest.fn();
jest.mock('react-router-dom', () => ({
	...jest.requireActual('react-router-dom'),
	useLocation: () => ({
		search: {
			slotId: '1',
		},
	}),
	useHistory: () => ({
		push: mockHistoryPush,
	}),
}));

const mockWorkers: IWorker[] = [
	{
		id: 1,
		name: 'Maxwell',
		rating: '4.1',
		isNew: false,
	},
	{
		id: 2,
		name: 'Melina',
		rating: '4.1',
		isNew: false,
	},
];
const mockSlot = {
	id: 1,
	localisedTime: '10:00',
	price: '$20',
};
jest.mock('../../../hooks/application', () => {
	return {
		useApplication: () => ({
			workers: mockWorkers,
			selectedSlot: mockSlot,
		}),
	};
});

const mockHandleError = jest.fn();
jest.mock('react-error-boundary', () => ({
	...jest.requireActual('react-error-boundary'),
	useErrorHandler: () => mockHandleError,
}));

jest.mock('uuid', () => ({
	...jest.requireActual('uuid'),
	v4: () => '1',
}));

const mockAddToCart = jest.fn();
jest.mock('../../../hooks/cart', () => {
	return {
		useCart: () => ({
			addToCart: mockAddToCart,
		}),
	};
});

describe('WorkersContainer', () => {
	const availableWorkers: IAvailableWorker[] = [
		{
			slotId: 1,
			availableWorkersIds: [1, 2, 3],
		},
	];

	beforeEach(() => {
		(WorkerService.getAvailableWorkers as jest.Mock).mockResolvedValue(
			availableWorkers
		);
	});

	it('should render correctly when service returns properties', async () => {
		const { container } = render(<WorkersContainer />);

		await waitFor(() => {
			expect(WorkerService.getAvailableWorkers).toHaveBeenCalledTimes(1);
			expect(WorkerService.getAvailableWorkers).toHaveBeenCalledWith(
				mockHandleError
			);
		});

		await waitFor(() => {
			expect(container.firstChild).toMatchSnapshot();
		});
	});

	it('should have a link with href to the home page', async () => {
		render(<WorkersContainer />);

		await waitFor(() => {
			expect(WorkerService.getAvailableWorkers).toHaveBeenCalledTimes(1);
			expect(WorkerService.getAvailableWorkers).toHaveBeenCalledWith(
				mockHandleError
			);
		});

		expect(screen.getByRole('link')).toHaveAttribute('href', HOME);
	});

	it('should filter the workers while the user types', async () => {
		render(<WorkersContainer />);

		await waitFor(() => {
			expect(WorkerService.getAvailableWorkers).toHaveBeenCalledTimes(1);
			expect(WorkerService.getAvailableWorkers).toHaveBeenCalledWith(
				mockHandleError
			);
		});

		const input = screen.getByPlaceholderText(/select a worker/i);
		expect(input).toBeInTheDocument();

		expect(screen.getAllByRole('option')).toHaveLength(2);

		userEvent.type(input, 'M');
		expect(screen.getAllByRole('option')).toHaveLength(2);

		userEvent.type(input, 'e');
		expect(screen.getAllByRole('option')).toHaveLength(1);

		await waitFor(() => {
			const bookButton = screen.getByRole('button', { name: /book/i });
			userEvent.click(bookButton);

			expect(mockAddToCart).toHaveBeenCalledTimes(1);
			expect(mockAddToCart).toHaveBeenCalledWith({
				worker: mockWorkers[1],
				id: '1',
				slot: mockSlot,
				price: 20,
				amount: 1,
			});
		});
	});
});
