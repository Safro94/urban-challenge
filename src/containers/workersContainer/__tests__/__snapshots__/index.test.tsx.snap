// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkersContainer should render correctly when service returns properties 1`] = `
<div
  class="sc-dQoVA cEIhtC"
>
  <div
    class="sc-bdfBQB dxzuBe"
  >
    <div
      class="sc-dlfnuX gTTmwj"
    >
      <label
        class="sc-eCstlR kAGrIR"
        for="downshift-0-input"
        id="downshift-0-label"
      >
        Select a worker
      </label>
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-owns="downshift-0-menu"
        class="sc-gsTEea fFdqYP"
        role="combobox"
      >
        <input
          aria-autocomplete="list"
          aria-controls="downshift-0-menu"
          aria-labelledby="downshift-0-label"
          autocomplete="off"
          class="sc-hKgJUU wJrVD"
          id="downshift-0-input"
          placeholder="Select a worker"
          value=""
        />
      </div>
    </div>
    <ul
      aria-labelledby="downshift-0-label"
      class="sc-kstqJO eGBVjW"
      id="downshift-0-menu"
      role="listbox"
    >
      <li
        aria-selected="false"
        class="sc-gKseQn dvkeCf"
        id="downshift-0-item-0"
        role="option"
      >
        <div
          class="sc-iBPTik cUqWwi"
        >
          <div
            class="sc-jrAFXE gZuALw"
          >
            <img
              alt="Worker 0"
              class="sc-iqHYmW hEHryu"
              src="therapist.jpeg"
            />
          </div>
          <div
            class="sc-pGacB bcJhql"
          >
            <div
              class="sc-kEjbQP iUkCGp"
            >
              <span
                class="sc-crrszt"
              >
                Name:
              </span>
              <span
                class="sc-crrszt"
              >
                Maxwell
              </span>
            </div>
            <div
              class="sc-kEjbQP iUkCGp"
            >
              <span
                class="sc-crrszt"
              >
                Rating:
              </span>
              <span
                class="sc-crrszt"
              >
                ⭐⭐⭐⭐
              </span>
            </div>
          </div>
          <div
            class="sc-fubCzh drXHda"
          >
            <button
              class="sc-jSgvzq hLltFr"
              type="button"
            >
              Book
            </button>
          </div>
        </div>
      </li>
      <li
        aria-selected="false"
        class="sc-gKseQn dvkeCf"
        id="downshift-0-item-1"
        role="option"
      >
        <div
          class="sc-iBPTik cUqWwi"
        >
          <div
            class="sc-jrAFXE gZuALw"
          >
            <img
              alt="Worker 1"
              class="sc-iqHYmW hEHryu"
              src="therapist.jpeg"
            />
          </div>
          <div
            class="sc-pGacB bcJhql"
          >
            <div
              class="sc-kEjbQP iUkCGp"
            >
              <span
                class="sc-crrszt"
              >
                Name:
              </span>
              <span
                class="sc-crrszt"
              >
                Melina
              </span>
            </div>
            <div
              class="sc-kEjbQP iUkCGp"
            >
              <span
                class="sc-crrszt"
              >
                Rating:
              </span>
              <span
                class="sc-crrszt"
              >
                ⭐⭐⭐⭐
              </span>
            </div>
          </div>
          <div
            class="sc-fubCzh drXHda"
          >
            <button
              class="sc-jSgvzq hLltFr"
              type="button"
            >
              Book
            </button>
          </div>
        </div>
      </li>
    </ul>
  </div>
  <div
    class="sc-bqyKOL iOnCHN"
  >
    <a
      class="sc-hBEYId hCthmQ"
      href="/"
    >
      Go back
    </a>
  </div>
</div>
`;
